java/util/src/main/java/com/google/protobuf/util/JsonFormat.java

```diff
@@ -1126,144 +1158,127 @@ public int compare(final Object o1, final Object o2) {
       generator.print("}");
     }
 
-    private void printSingleFieldValue(FieldDescriptor field, Object value) throws IOException {
+    private void printKey(
+        final FieldDescriptor field, final Object value)
+        throws IOException {
+      printSingleFieldValue(field, value, true);
+    }
+
+    private void printSingleFieldValue(
+        final FieldDescriptor field, final Object value)
+        throws IOException {
       printSingleFieldValue(field, value, false);
     }
 
     /**
      * Prints a field's value in the proto3 JSON format.
-     *
-     * @param alwaysWithQuotes whether to always add double-quotes to primitive
-     *        types
      */
     private void printSingleFieldValue(
         final FieldDescriptor field, final Object value, boolean alwaysWithQuotes)
         throws IOException {
+      boolean quoteable = false;
+      String stringValue = null;
+
       switch (field.getType()) {
         case INT32:
         case SINT32:
         case SFIXED32:
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
-          generator.print(((Integer) value).toString());
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
+          stringValue = ((Integer) value).toString();
           break;
 
         case INT64:
         case SINT64:
         case SFIXED64:
-          generator.print("\"" + ((Long) value).toString() + "\"");
+          quoteable = true;
+          stringValue = ((Long) value).toString();
           break;
 
         case BOOL:
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
-          if (((Boolean) value).booleanValue()) {
-            generator.print("true");
-          } else {
-            generator.print("false");
-          }
-          if (alwaysWithQuotes) {
-            generator.print("\"");
-          }
+          stringValue = (Boolean) value ? "true" : "false";
           break;
 
         case FLOAT:
           Float floatValue = (Float) value;
           if (floatValue.isNaN()) {
-            generator.print("\"NaN\"");
+            stringValue = "\"NaN\"";
           } else if (floatValue.isInfinite()) {
             if (floatValue < 0) {
-              generator.print("\"-Infinity\"");
+              stringValue = "\"-Infinity\"";
             } else {
-              generator.print("\"Infinity\"");
+              stringValue = "\"Infinity\"";
             }
           } else {
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
-            generator.print(floatValue.toString());
-            if (alwaysWithQuotes) {
-              generator.print("\"");
-            }
+            stringValue = floatValue.toString();
           }
           break;
 
         case DOUBLE:
           Double doubleValue = (Double) value;
           if (doubleValue.isNaN()) {
-            generator.print("\"NaN\"");
+            alwaysWithQuotes = true;
```

Comment:
Whatever logic strategy we settle on, FLOAT and DOUBLE should have equivalent implementation.